File: aoc23/day3/main.go
package main

import (
	"fmt"
	"os"
	"strings"
	"unicode"
)

func main() {
	data, _ := os.ReadFile("input.txt")
	fmt.Println(part2(string(data)))
}

func solve(data string) int {
	lines := strings.Split(data, "\n")
	lines = lines[0 : len(lines)-1]
	// fmt.Println(lines, len(lines))
	grid := make([][]byte, len(lines))
	for i, line := range lines {
		grid[i] = []byte(line)
	}

	charPresent := false
	builder, total := 0, 0

	h, w := len(grid), len(grid[0])

	for i := 0; i < h; i++ {
		// fmt.Printf("Parsing %s\n", string(grid[i]))
		for j := 0; j < w; j++ {
			if unicode.IsDigit(rune(grid[i][j])) {
				builder = builder*10 + int(grid[i][j]-'0')
				if !charPresent {
					charPresent = isSymbolInSorroundings(grid, i, j, h, w)
				}
			} else {
				if builder != 0 && charPresent {
					total += builder
					// println("adding", builder)
				}

				builder = 0
				charPresent = false
			}
			// println(builder, charPresent)
		}

		if builder != 0 && charPresent {
			// println("adding", builder)
			total += builder
		}
		builder = 0
		charPresent = false
	}

	return total
}

var directionsX []int = []int{-1, 0, 1}
var directionsY []int = []int{-1, 0, 1}

func isSymbol(s byte) bool {
	return !unicode.IsDigit(rune(s)) && s != '.'
}

func isSymbolInSorroundings(grid [][]byte, i int, j int, h int, w int) bool {
	for _, dx := range directionsX {
		for _, dy := range directionsY {
			if dx == 0 && dy == 0 {
				continue
			}

			i_ := i + dx
			j_ := j + dy

			if i_ < h && j_ < w && i_ >= 0 && j_ >= 0 {
				if isSymbol(grid[i_][j_]) {
					return true
				}
			}
		}
	}

	return false
}

func part2(data string) int {
	lines := strings.Split(data, "\n")
	lines = lines[0 : len(lines)-1]
	// fmt.Println(lines, len(lines))
	grid := make([][]byte, len(lines))
	for i, line := range lines {
		grid[i] = []byte(line)
	}

	var total int = 0
	h, w := len(grid), len(grid[0])

	for i := 0; i < h; i++ {
		// fmt.Printf("Parsing %s\n", string(grid[i]))
		for j := 0; j < w; j++ {
		if grid[i][j] == '*' {
				neighbors := getNeighboringNums(grid, i, j, h, w)
				fmt.Println(neighbors)
				if len(neighbors) == 2 {
					total += neighbors[0] * neighbors[1]
				}
			}
		}
	}

	return total
}

func getNeighboringNums(grid [][]byte, i int, j int, h int, w int) []int {
	nums := []int{}

	sides := [][][]int{
		{
			{-1, -1}, {-1, 0}, {-1, 1},
		},
		{
			{0, -1},
		},
		{
			{0, 1},
		},
		{
			{1, -1}, {1, 0}, {1, 1},
		},
	}

	for _, side := range sides {
		continous := false
		for _, points := range side {
			di, dj := points[0], points[1]
			i_ := i + di
			j_ := j + dj

			if inBounds(i_, j_, h, w) && unicode.IsDigit(rune(grid[i_][j_])) {
				if !continous {
					nums = append(nums, propNum(grid, i_, j_, h, w))
				}
				continous = true
			} else {
				continous = false
			}
		}
	}

	return nums
}

func inBounds(i, j, h, w int) bool {
	return i < h && j < w && i >= 0 && j >= 0
}

func propNum(grid [][]byte, i, j, h, w int) int {
	number := 0
	iter := j
	for ; inBounds(i, iter, h, w) && unicode.IsDigit(rune(grid[i][iter])); iter-- {
	}

	iter++ // back to start of number
	for ; inBounds(i, iter, h, w) && unicode.IsDigit(rune(grid[i][iter])); iter++ {
		number = number*10 + int(grid[i][iter]-'0')
	}

	return number
}

File: aoc23/day3/src/main.rs
use std::{fs, io::Read};

fn main() {
    let mut file = fs::File::open("input.txt").unwrap();
    let mut data = String::new();
    let _ = file.read_to_string(&mut data);

    let output = solve(&data);
    println!("{output}");
}

fn solve(data: &str) -> i32 {
    let grid: Vec<Vec<char>> = data.lines().map(|s| s.chars().collect()).collect();
    let h = grid.len();
    let w = grid[0].len();

    let mut builder = 0;
    let mut char_found = false;
    let mut total = 0;
    for i in 0..h {
        for j in 0..w {
            if grid[i][j].is_digit(10) {
                let ch = grid[i][j] as i32 - '0' as i32;
                builder = builder * 10 + ch;

                if i > 0 {
                    if j > 0 {
                        let tl = grid[i - 1][j - 1];
                        if check_symbol(tl) {
                            char_found = true;
                        }
                        if check_symbol(grid[i][j - 1]) {
                            char_found = true;
                        }
                    }

                    if j + 1 < w {
                        let tr = grid[i - 1][j + 1];
                        if check_symbol(tr) {
                            char_found = true;
                        }
                        if check_symbol(grid[i][j + 1]) {
                            char_found = true;
                        }
                    }
                }

                if i + 1 < h {
                    if j > 0 {
                        let bl = grid[i + 1][j - 1];
                        if check_symbol(bl) {
                            char_found = true;
                        }
                        if check_symbol(grid[i][j - 1]) {
                            char_found = true;
                        }
                    }

                    if j + 1 < w {
                        let br = grid[i + 1][j + 1];
                        if check_symbol(br) {
                            char_found = true;
                        }
                        if check_symbol(grid[i][j + 1]) {
                            char_found = true;
                        }
                    }
                }
            } else {
                if char_found {
                    total += builder;
                }
                if builder > 0 && !char_found {
                    println!("builder: {builder}");
                    if i > 0 {
                        println!("{}", grid[i - 1].iter().collect::<String>());
                    }
                    println!("{}", grid[i].iter().collect::<String>());
                    if i < h - 1 {
                        println!("{}", grid[i + 1].iter().collect::<String>());
                    }

                    println!("\n\n");
                }
                char_found = false;
                builder = 0;
            }
        }

        if char_found {
            total += builder;
            builder = 0;
            char_found = false;
        }
    }

    return total;
}

fn check_symbol(c: char) -> bool {
    // return c == '#' || c == '@' || c == '$' || c == '%' || c == '&' || c == '*';
    return !c.is_digit(10) && c != '.';
}
